### PR TITLE
[Enhancement] cloud native pk table support size tiered compaction policy

### PR DESCRIPTION
**Pull request by luohaha**
## Why I'm doing:
This PR Introduces Size Tiered Compaction Policy to cloud native primary key model, to reduce write amplification in small batch data ingestion and update scenarios.

## What I'm doing:
There are several changes in this PR:
1. Refactor code, move PrimaryCompactionPolicy to separate file `primary_key_compaction_policy.cpp`.
2. When pick rowsets, we separate rowsets into different rowset levels by data size. And then pick the level with largest score, then we will use rowsets in this level as compaction input rowsets.
3. Add config `enable_pk_size_tiered_compaction_strategy` to control new policy.

#### Test Result:
We make a test under two kinds of data size:
1. 1G
2. 2.5G

And will test three different ingestion scenarios:
1. Ingestion from empty, insert 1/100 of total data each time.
2. upsert 1/100 of total data each time.
3. upsert 1/10000 of total data each time.

* Without Size Tiered Compaction Policy

  | Ingestion size | total write size | Write amplification
-- | -- | -- | --
1G + insert + 1/100 | 1206701190 | 4170126834 | 3.45
1G + upsert + 1/100 | 1206466991 | 7350584227 | 6.1
1G + upsert + 1/10000 | 90488370 | 10409035154 | 115
2.5G + insert + 1/100 | 2515987204 | 8536016843 | 3.39
2.5G + upsert + 1/100 | 2515986927 | 12324324616 | 4.89
2.5G + upsert + 1/10000 | 334647851 | 20597638783 | 61.5

* With Size Tiered Compaction Policy

  | Ingestion size | total write size | Write amplification
-- | -- | -- | --
1G + insert + 1/100 | 1206701163 | 3077673322 | 2.55
1G + upsert + 1/100 | 1206701436 | 5796920239 | 4.8
1G + upsert + 1/10000 | 90316696 | 564930454 | 6.25
2.5G + insert + 1/100 | 2515987622 | 6780197910 | 2.69
2.5G + upsert + 1/100 | 2465693778 | 9691248064 | 3.9
2.5G + upsert + 1/10000 | 333182197 | 3472062137 | 10.42

We can find out that write amplification reduce when using STCP.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by sduzh** [comment]
@starrocks-xupeng PTAL

File: be/src/storage/lake/primary_key_compaction_policy.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+StatusOr<std::vector<RowsetPtr>> PrimaryCompactionPolicy::pick_rowsets(
+        const std::shared_ptr<const TabletMetadataPB>& tablet_metadata, bool calc_score, std::vector<bool>* has_dels) {
+    std::vector<RowsetPtr> input_rowsets;
+    UpdateManager* mgr = _tablet_mgr->update_mgr();
+    std::vector<RowsetCandidate> rowset_vec;
+    const auto tablet_id = tablet_metadata->id();
+    const auto tablet_version = tablet_metadata->version();
+    const int64_t compaction_data_size_threshold =
+            static_cast<int64_t>((double)_get_data_size(tablet_metadata) * config::update_compaction_ratio_threshold);
+    // 1. generate rowset condidate vector
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/primary_key_compaction_policy.cpp]: Done

File: be/src/storage/lake/primary_key_compaction_policy.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            std::max(config::update_compaction_result_bytes, compaction_data_size_threshold)) {
+            break;
+        }
+        // If calc_score is true, we skip `config::lake_pk_compaction_max_input_rowsets` check,
+        // because `config::lake_pk_compaction_max_input_rowsets` is only used to limit the number
+        // of rowsets for real compaction merges
+        if (!calc_score && rowset_indexes.size() >= config::lake_pk_compaction_max_input_rowsets) {
+            break;
+        }
+        pick_level_ptr->rowsets.pop();
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/primary_key_compaction_policy.cpp:129]: It's on purpose, to avoid copy rowset condidate.

**Comment by luohaha** [comment]
> It seems that we should implement `pick_rowset_indexes` like #41081.

Already rebase.

File: be/src/storage/lake/primary_key_compaction_policy.h [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                     << " Dels: " << rowset.rowset_meta_ptr->num_dels() << " Score: " << rowset.score << "] ";
+        }
+        if (print_log) {
+            VLOG(2) << "PKSizeTieredLevel " << debug_ss.str();
+        }
+    }
+
+    // Merge another level's top rowset
+    void merge_top(const PKSizeTieredLevel& other) {
+        if (!other.rowsets.empty()) {
```
Comment (wyb) [inline be/src/storage/lake/primary_key_compaction_policy.h:99]: DCHECK(other.rowsets.size() == 1);

**Review by TszKitLo40 — APPROVED** [review]

File: be/src/storage/lake/primary_key_compaction_policy.h (hunk 3 above)
Comment (luohaha) [reply be/src/storage/lake/primary_key_compaction_policy.h:100]: Done

**Review by wyb — APPROVED** [review]

**Review by kevincai — APPROVED** [review]

**Comment by evelynzhaojie** [comment]
@luohaha pls add the meaning of the new parameter and the default value change to doc "BE configuration" https://github.com/StarRocks/starrocks/blob/main/docs/zh/administration/BE_configuration.md

**Pull request merged**